### PR TITLE
Add HTTP types adapter for SwiftNIO

### PR DESCRIPTION
**Pull request by guoye-zhang**
Add [swift-http-types](https://github.com/apple/swift-http-types) adapters that works with SwiftNIO HTTP protocols.

### Motivation:

SwiftNIO Channel implementations should be able to use the shared HTTP currency types.

### Modifications:

3 new targets:

* `NIOHTTPTypes`: version independent data types
* `NIOHTTPTypesHTTP1`: convertors between http1 types and new http types
* `NIOHTTPTypesHTTP2`: convertors between http2 types and new http types

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-server-bot add to allowlist

**Comment by tomerd** [comment]
@guoye-zhang I think this is missing the dependency?

**Comment by guoye-zhang** [comment]
Oh we need to reduce the swift version of swift-http-types to 5.6

**Comment by guoye-zhang** [comment]
I updated the Package.swift to only depend on swift-http-types on 5.8 or later since it depends on a few new language features.

**Comment by tomerd** [comment]
cc @Lukasa

**Review by tomerd — APPROVED** [review]

File: Package.swift
Comment (FranzBusch) [inline Package.swift:149]: We can't merge this until we tag a `1.0.0` in swift-http-types

File: Sources/NIOHTTPTypes/NIOHTTPTypes.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypes/NIOHTTPTypes.swift:24]:
```suggestion
public enum HTTPTypePart<Head: Equatable, Body: Equatable> {
```

File: Sources/NIOHTTPTypes/NIOHTTPTypes.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypes/NIOHTTPTypes.swift:35]: Can we get rid of `IOData` here the problem is that `IOData` is not `Sendable`

File: Sources/NIOHTTPTypes/NIOHTTPTypes.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypes/NIOHTTPTypes.swift:44]: Same here

File: Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift:32]:
```suggestion
        switch self.unwrapInboundIn(data) {
```

File: Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift:35]:
```suggestion
                try context.fireChannelRead(self.wrapInboundOut(.head(head.newResponse)))
```

File: Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift:42]:
```suggestion
            context.fireChannelRead(self.wrapInboundOut(.body(body)))
        case .end(let trailers):
            context.fireChannelRead(self.wrapInboundOut(.end(trailers?.newFields(splitCookie: false))))
```

Comment (FranzBusch) [reply Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift:42]: Can you add `self.` in the whole diff please.

File: Sources/NIOHTTPTypes/NIOHTTPTypes.swift
Comment (FranzBusch) [reply Sources/NIOHTTPTypes/NIOHTTPTypes.swift:35]: Maybe we can get rid off the two different client/server types all together then cc @Lukasa

File: Package.swift
Comment (FranzBusch) [inline Package.swift:140]: This is a bit unfortunate. What 5.8 language features are we using that we can't use in 5.7? Is it only the `some` keyword? If so it would be great if we could just write this differently and make it support 5.7

File: Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift
Comment (Lukasa) [inline Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift:126]: As a general style nit we prefer to use `&&` where it is possible to do so.

File: Sources/NIOHTTPTypes/NIOHTTPTypes.swift
Comment (guoye-zhang) [reply Sources/NIOHTTPTypes/NIOHTTPTypes.swift:35]: Is it OK to synchronously read `FileRegion` into `ByteBuffer` when converting an `HTTPPart` into an `HTTPTypePart`?

Comment (guoye-zhang) [reply Sources/NIOHTTPTypes/NIOHTTPTypes.swift:35]: I'll use a `fatalError` for now

File: Package.swift
Comment (guoye-zhang) [reply Package.swift:140]: 5.8 introduced a few buffer methods (SE-0370) which we are using. I could rewrite it to use memcpy

Comment (guoye-zhang) [reply Package.swift:140]: https://github.com/apple/swift-http-types/pull/2

File: Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift
Comment (guoye-zhang) [reply Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift]: This was actually changed to `,` by swiftformat. I changed it back.

File: Sources/NIOHTTPTypes/NIOHTTPTypes.swift
Comment (Lukasa) [reply Sources/NIOHTTPTypes/NIOHTTPTypes.swift]: Nope, unfortunately it isn't: doing that I/O on NIO's event loops isn't acceptable.

File: Sources/NIOHTTPTypesHTTP1/HTTPToHTTP1Codec.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypesHTTP1/HTTPToHTTP1Codec.swift:111]: Same here

File: Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift:172]: We normally spell this as an `init` on `HTTPRequest` that takes an `HTTPRequestHead` and the two other parameters.

File: Sources/NIOHTTPTypesHTTP2/HTTPTypeConversion.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypesHTTP2/HTTPTypeConversion.swift:44]: Same here. Normally spelled with an `init` on `HTTPField.DynamicTableIndexingStrategy`

File: Sources/NIOHTTPTypesHTTP2/HTTPTypeConversion.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypesHTTP2/HTTPTypeConversion.swift:101]: Same here.

File: Sources/NIOHTTPTypesHTTP2/HTTPTypeConversion.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypesHTTP2/HTTPTypeConversion.swift:196]: and here

File: Sources/NIOHTTPTypesHTTP2/HTTPTypeConversion.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypesHTTP2/HTTPTypeConversion.swift:248]: and here

**Review by FranzBusch — COMMENTED** [review]
Overall I think this is mostly ready. I left a couple of comments inline.

File: Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift
Comment (guoye-zhang) [reply Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift:53]: Do we need both `fireErrorCaught` and failing the promise?

Comment (guoye-zhang) [reply Sources/NIOHTTPTypesHTTP1/HTTP1ToHTTPCodec.swift:53]: Oh I see other places doing both

File: Sources/NIOHTTPTypes/NIOHTTPTypes.swift
Comment (guoye-zhang) [reply Sources/NIOHTTPTypes/NIOHTTPTypes.swift]: I made them separate enums, but requests can support trailers as far as I know

File: Package.swift
Comment (FranzBusch) [reply Package.swift]: Interesting what do you need from the patch release? @Lukasa I think we can bump to the patch release right away. At least there is prior art in doing this AFAIK

File: Sources/NIOHTTPTypes/NIOHTTPTypes.swift
Comment (FranzBusch) [reply Sources/NIOHTTPTypes/NIOHTTPTypes.swift]: Yeah I learned this today as well after reading the RFC again. Requests can indeed container trailers.

File: Package.swift
Comment (guoye-zhang) [reply Package.swift]: I think some String API was added in 5.7.1: https://github.com/apple/swift-http-types/issues/20

File: Package.swift
Comment (FranzBusch) [inline Package.swift:140]: Just drop this conditional completely. We are going to merge this PR after we drop 5.6 and then we will just make `5.7.1` or bottom version.

File: Sources/NIOHTTPTypes/NIOHTTPTypes.swift
Comment (FranzBusch) [inline Sources/NIOHTTPTypes/NIOHTTPTypes.swift:38]: NIT: I am wondering if we need the `HTTPType` prefix or can just go with `HTTPRequestPart` and `HTTPResponsePart`. Those names haven't been used yet in the NIO ecosystem from what I can tell.

Comment (guoye-zhang) [reply Sources/NIOHTTPTypes/NIOHTTPTypes.swift]: Updated

File: Package.swift
Comment (guoye-zhang) [reply Package.swift]: Fixed

**Review by FranzBusch — APPROVED** [review]
LGTM! @Lukasa mind giving this a look as well

**Comment by guoye-zhang** [comment]
Should I bump the Swift version to 5.7.1 in this PR?

**Comment by FranzBusch** [comment]
> Should I bump the Swift version to 5.7.1 in this PR?

Yes let's do that.

**Comment by guoye-zhang** [comment]
There are still a few Dockerfiles mentioning 5.7, but those should mean latest 5.7.x? Do I need to change them?

**Comment by Lukasa** [comment]
They mean latest 5.7.x. That's typically fine, 5.7 just means "floating" in most cases.

File: Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift
Comment (guoye-zhang) [reply Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift:128]: RFC7230 has this sentence which is no longer present in RFC9112:

>    Since the Host field-value is critical information for handling a
>    request, a user agent SHOULD generate Host as the first header field
>    following the request-line.

I guess we can't assume that it is always first. HTTPTypes does not preserve the order of pseudo header fields, so we will lose the order of Host (which might be fine?)

File: Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift
Comment (guoye-zhang) [reply Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift:132]: Cookie splitting is a performance optimization and it isn't done today with HTTP2FramePayloadToHTTP1ClientCodec. URLSession only started doing it in iOS 14. It's not ideal, but it's not worth spending a lot of effort into I think

File: Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift
Comment (guoye-zhang) [reply Sources/NIOHTTPTypesHTTP1/HTTPTypeConversion.swift]: Fixed in 6da1b1a

**Review by Lukasa — APPROVED** [review]
Nice one, thanks @guoye-zhang!

**Pull request merged**
